Integrate line detection model into Transkribus engine

Parthiv-M opened the pull request:
Add API routes, helper methods and UI components to integrate line detection models into the Transkribus OCR engine

Bug: [T340837](https://phabricator.wikimedia.org/T340837)

Commit:
Integrate line detection model into Transkribus engine

Add API routes, helper methods and UI components
to integrate line detection models into the
Transkribus OCR engine

Bug: T340837

samwilson commented on src/Controller/OcrController.php:
```diff
@@ -262,6 +279,29 @@ public function apiAvailableLangsAction(): JsonResponse {
 
+	/**
+	 * phpcs:disable MediaWiki.Commenting.FunctionAnnotations.UnrecognizedAnnotation
+	 * @Route("/api/available_line_ids", name="apiLineIds", methods={"GET"})
```
As line detection models only ever apply to Transkribus, perhaps this API endpoint could be simplified to e.g. `/api/transkribus/line_ids` and not have the `engine` parameter?

It could also have a `lang` parameter, so it'd only return line IDs for a specific language model, but that's probably not worth it given it's such a small result set.

Parthiv-M replied:
I just noticed, the `?engine` query is required by the [`setup()` method](https://github.com/wikimedia/wikimedia-ocr/blob/79dc7d0598c33bc18932e78edcccea6e71d34028/src/Controller/OcrController.php#L91) to set the OCR engine for all further operations.

samwilson replied:
Maybe it could set `static::$params['engine']` in this method? As it stands, people could pass e.g. `/api/available_line_ids?engine=google`, which wouldn't really make any sense.

Parthiv-M replied:
Yupp, doing that

samwilson commented on src/Engine/TranskribusEngine.php:
```diff
@@ -12,6 +12,12 @@ class TranskribusEngine extends EngineBase {
+	protected $lineId;
+
+	/** @var int Default value for line detection model ID to be used by Transkribus */
+	public const DEFAULT_LINEID = 0;
```
Does `0` have a specific meaning for this in the Transkribus API? If not, it might be neater to make the default `null` and check instead for that above.

Parthiv-M replied:
`null` should be fine

samwilson commented on templates/_transkribus_options.html.twig:
```diff
@@ -0,0 +1,23 @@
+<fieldset id="transkribus-line-ids" class="engine-options {% if engine != 'transkribus' %}hidden{% endif %}">
```
The ID of this element should be `transkribus-options` (in line with the tesseract-options element) but that's already in use in `_transkribus_help.html.twig`. I'd suggest renaming that, so that this gets its rightful name. That will also simplify the JS switching logic.

Parthiv-M replied:
Got it

samwilson commented on assets/app.js:
```diff
@@ -43,6 +44,30 @@ function updateSelect2Options(engine)
+ * the line detection model IDs available for the Transkribus engine
+ */
+function updateLineModelOptions () {
+    $.getJSON('/api/available_line_ids?engine=transkribus').then(response => {
```
Does this need to be re-fetched every time the engine selection is changed? It currently returns the same thing every time, so could probably just be done once on page load.

It could also perhaps be sending the language code and so only showing the valid line IDs for the currently-selected language.

samwilson commented on templates/_transkribus_options.html.twig:
```diff
@@ -0,0 +1,23 @@
+            data-placeholder="Choose a line detection model"
+        >
+            <option value="0" {% if lineId == 0 %}selected{% endif %}>None</option>
+            <option value="49272" {% if lineId == 49272 %}selected{% endif %}>Default line detection model</option>
```
Here and above in `app.js` it's adding these two options, but I'm confused about what the difference is between *not* sending a line ID and sending what we know to be the *default*. This looks like it'd be fine for now (although unnecessary), but if the default ever changes then this will be incorrect. Alternatively, 49272 is not applied unless we send it, but in that case we shouldn't be calling it 'default'.

Also, the text here and for the placeholder should be translated.

Parthiv-M replied:
It's more of a semantic error on my part 😅. No model being selected implies we let Transkribus choose the model automatically, whereas using the default model (49272) implies using the _public model that works best in most cases_, as suggested by Transkribus. 

[According to Transkribus](https://transkribus.eu/TrpServer/rest/models/layout/49272), model 49272 is named _Mixed Line Orientation_. Can we go ahead and use the same name?

samwilson replied:
Ah that makes sense, and yep that name sounds fine (although I wonder how it'll translate… it's no worse than the PSM names though).

Parthiv-M replied:
Okayy

samwilson commented on src/Controller/OcrController.php:
```diff
@@ -165,6 +175,13 @@ public function homeAction(): Response {
+
 		// Intution::listToText() isn't available via Twig, and we only want to do this for the view and not the API.
 		static::$params['image_hosts'] = $this->intuition->listToText( static::$params['image_hosts'] );
 
```
Below here in `apiAction()` shouldn't there be a new parameter of `line_id`?

samwilson commented:
Also, I'm getting a 403 response from Transkribus with the wikimedia account: "Your account does not have access to text recognition model with ID = 45612". How do I fix that? Do you have it working correctly?

Parthiv-M commented:
> Also, I'm getting a 403 response from Transkribus with the wikimedia account: "Your account does not have access to text recognition model with ID = 45612". How do I fix that? Do you have it working correctly?

I believe this 403 response is because the Balinese line detection model is a private model. We might need to reach out to Transkribus regarding that.

samwilson commented on src/Engine/EngineBase.php:
```diff
@@ -143,6 +148,46 @@ public function getValidLangs( bool $withNames = false ): array {
+	 * @param bool $onlyLineIdLangs Whether to return only the line detection model IDs lang codes
+	 * @return string[] Line detection model lang codes or model IDs or model ID names
+	 */
+	public function getValidLineIds( bool $onlyLineIds = false, bool $onlyLineIdLangs = false ): array {
```
I'm not sure Line IDs should be part of the EngineBase; they're an attribute just of Transkribus, so shouldn't be at this higher level.

samwilson commented on src/Engine/EngineBase.php:
```diff
@@ -154,10 +199,19 @@ public function getLangName( ?string $lang = null ): string {
+	}
+
 	/**
 	 * Transform the given ISO 639-1 codes into the language codes needed by this type of Engine.
```
This comment, and perhaps the method name as well, should probably be updated to reflect that the return value is 'language model information' or something, rather than just "the language codes".

Parthiv-M replied:
I'm not sure I get you on this. This method just returns the name of the model from the array `LINE_ID_MODEL_NAMES`.

Commit:
Update code structure, modify API route,
and make other minor enhancements to logic
for line detection model selection for Transkribus

Commit:
Modify parameters for AppExtensionTest test case

Commit:
Add missing import statements

Commit:
Replace variable with Intuition class instance

Commit:
Update parameters in AppExtension test case

samwilson commented on src/Controller/OcrController.php:
```diff
@@ -165,6 +175,15 @@ public function homeAction(): Response {
 
+		if ( static::$params['engine'] === 'transkribus' ) {
+			// Pre-supply the available line ids for autocompletion in the form.
+			static::$params['available_line_ids'] = $this->engine->getValidLineIds( true, false );
```
This array key also needs to be set in `OcrController::$params`, otherwise it throws an error when loading without `?engine=transkribus`:

> Variable "available_line_ids" does not exist.
> in templates/_transkribus_options.html.twig (line 12)

samwilson commented on src/Controller/OcrController.php:
```diff
@@ -60,6 +60,7 @@ class OcrController extends AbstractController {
 		'langs' => [],
 		'psm' => TesseractEngine::DEFAULT_PSM,
 		'crop' => [],
+		'lineId' => TranskribusEngine::DEFAULT_LINEID
```
Nitpick: it doesn't matter at all, but I think URL & API component names should either be camelCase or snake_case, but not both. It looks like we've got more of the latter so far (although most are single words so aren't either), so I suggest `line_id` here.

samwilson commented on src/Controller/OcrController.php:
```diff
@@ -262,6 +287,22 @@ public function apiAvailableLangsAction(): JsonResponse {
+		$this->request->query->set( 'engine', 'transkribus' );
+		$this->setup();
+		return $this->getApiResponse( [
+			'engine' => 'transkribus',
```
As `transkribus` is now in the API endpoint URL, there's probably no need to also list it in the returned data.

samwilson reviewed (commented):
Looks good, works in my tests. Just one bug with there being no default value.

Commit:
Fix minor bugs
